src/embind/embind.js

```diff
@@ -1005,24 +1009,38 @@ var LibraryEmbind = {
 
       var rv = cppInvokerFunc.apply(null, invokerFuncArgs);
 
-      if (needsDestructorStack) {
-        runDestructors(destructors);
-      } else {
-        for (var i = isClassMethodFunc ? 1 : 2; i < argTypes.length; i++) {
-          var param = i === 1 ? thisWired : argsWired[i - 2];
-          if (argTypes[i].destructorFunction !== null) {
-            argTypes[i].destructorFunction(param);
+      function onDone(rv) {
+        if (needsDestructorStack) {
+          runDestructors(destructors);
+        } else {
+          for (var i = isClassMethodFunc ? 1 : 2; i < argTypes.length; i++) {
+            var param = i === 1 ? thisWired : argsWired[i - 2];
+            if (argTypes[i].destructorFunction !== null) {
+              argTypes[i].destructorFunction(param);
+            }
           }
         }
+
+  #if EMSCRIPTEN_TRACING
+        Module.emscripten_trace_exit_context();
+  #endif
+
+        if (returns) {
+          return argTypes[0].fromWireType(rv);
+        }
       }
 
-#if EMSCRIPTEN_TRACING
-      Module.emscripten_trace_exit_context();
+#if ASYNCIFY
+      if (Asyncify.currData) {
```

Comment:
What is this checking specifically? It would probably be better to do something explicit like
```
Asyncify.state === Asyncify.State.Unwinding
```

Reply:
I essentially followed the example of `async:true` handling in `ccall` here that also checks `Asyncify.currData`: https://github.com/emscripten-core/emscripten/blob/e6f4d495668b7bfea079e03779280ae26f413186/src/preamble.js#L134

I guess the `state` check would be better yeah... need to double-check that it's doing the same thing.

Reply:
Ah, good point, we should fix `ccall` as well. I'll look into that.

Reply:
Oh, actually I was wrong. The state we care about here is after unwinding and before rewinding. That's not an asyncify state. And it is indicated as currData not being null. So sgtm as is.

Reply:
Ah, thanks for checking.